### PR TITLE
Add more information about DI to V8 upgrade guide

### PR DESCRIPTION
**Pull request by timbussmann**
the migration guide currently doesn't include ninject. Should we add a section stating that ninject doesn't support the MS DI abstraction atm?

### CODE REVIEW

**Comment by timbussmann** [comment]
@DavidBoike do you have an idea how I need to fix that anchor for the migration section on the same document?

File: nservicebus/upgrades/7to8/dependency-injection.md [hunk 1]
```diff
@@ -0,0 +1,102 @@
+---
+title: Dependency Injection Changes
```
Comment (mauroservienti) [inline nservicebus/upgrades/7to8/dependency-injection.md:2]:
```suggestion
title: Dependency Injection changes
```

File: nservicebus/upgrades/7to8/dependency-injection.md [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+summary: Dependency injection changes from NServiceBus 7 to 8.
```
Comment (mauroservienti) [inline nservicebus/upgrades/7to8/dependency-injection.md:3]:
```suggestion
summary: Dependency Injection changes from NServiceBus 7 to 8.
```

File: nservicebus/upgrades/7to8/dependency-injection.md [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+
+Property injection is not covered by `Microsoft.Extensions.DependencyInjection.Abstractions`, therefore the NServiceBus default dependency injection container does not support property injection anymore. Property injection might be supported by third party containers that can be enabled using the [externally managed container mode](/nservicebus/dependency-injection/#externally-managed-mode).
+
+## UseContainer is deprecated
+
+The `UseContainer` API to integrate third party containers with NServiceBus has been removed as it does not align with the `Microsoft.Extensions.DependencyInjection.Abstractions` model. To use a custom dependency injection container with NServiceBus, use the [externally managed container mode](/nservicebus/dependency-injection/#externally-managed-mode).
+
+## Externally managed container mode
+
+The externally managed container mode allows to integrate third party dependency injection containers with NServiceBus that conform to the `Microsoft.Extensions.DependencyInjection.Abstractions` model via the `EndpointWithExternallyManagedContainer.Create` API. See the following sections for examples using common DI containers.
```
Comment (mauroservienti) [inline nservicebus/upgrades/7to8/dependency-injection.md:34]:
```suggestion
The externally managed container mode allows integrating third-party dependency injection containers, that conform to the `Microsoft.Extensions.DependencyInjection.Abstractions` model, with NServiceBus via the `EndpointWithExternallyManagedContainer.Create` API. See the following sections for examples of using common DI containers.
```

File: nservicebus/upgrades/7to8/dependency-injection.md [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+
+```csharp
+var serviceCollection = new ServiceCollection();
+var startableEndpoint = EndpointWithExternallyManagedContainer.Create(endpointConfiguration, serviceCollection);
+
+var containerBuilder = new ContainerBuilder();
+containerBuilder.Populate(serviceCollection);
+var autofacContainer = containerBuilder.Build();
+var endpointInstance = await startableEndpoint.Start(new AutofacServiceProvider(autofacContainer));
+```
```
Comment (mauroservienti) [inline nservicebus/upgrades/7to8/dependency-injection.md:61]: I'd configure Autofac first and then create the startable endpoint.

File: nservicebus/upgrades/7to8/dependency-injection.md [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+The following snippet configures NServiceBus to use Castle Windsor as its dependency injection container. This requires the `Castle.Windsor.MsDependencyInjection` NuGet package.
+
+```csharp
+var serviceCollection = new ServiceCollection();
+var startableEndpoint = EndpointWithExternallyManagedContainer.Create(endpointConfiguration, serviceCollection);
+
+var container = new WindsorContainer();
+var serviceProvider = WindsorRegistrationHelper.CreateServiceProvider(container, serviceCollection);
+var endpointInstance = await startableEndpoint.Start(serviceProvider);
+```
```
Comment (mauroservienti) [inline nservicebus/upgrades/7to8/dependency-injection.md:89]: Same as above

File: nservicebus/upgrades/7to8/dependency-injection.md [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+The following snippet configures NServiceBus to use Unity as its dependency injection container. This requires the `Castle.Windsor.MsDependencyInjection` NuGet package.
+
+```csharp
+var serviceCollection = new ServiceCollection();
+var startableEndpoint = EndpointWithExternallyManagedContainer.Create(endpointConfiguration, serviceCollection);
+
+var container = new WindsorContainer();
+var serviceProvider = WindsorRegistrationHelper.CreateServiceProvider(container, serviceCollection);
+var endpointInstance = await startableEndpoint.Start(serviceProvider);
+```
```
Comment (mauroservienti) [inline nservicebus/upgrades/7to8/dependency-injection.md:102]: Same as above

File: nservicebus/upgrades/7to8/dependency-injection.md [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+```csharp
+var serviceCollection = new ServiceCollection();
+var startableEndpoint = EndpointWithExternallyManagedContainer.Create(endpointConfiguration, serviceCollection);
+
+var container = new Container();
+container.Populate(serviceCollection);
+var endpointInstance = await startableEndpoint.Start(container.GetInstance<IServiceProvider>());
+```
+
+Note: StructureMap has been sunsetted. The maintainers recommend to use [Lamar](https://jasperfx.github.io/lamar/) instead.
```
Comment (mauroservienti) [inline nservicebus/upgrades/7to8/dependency-injection.md:76]:
```suggestion

```

File: nservicebus/upgrades/7to8/dependency-injection.md [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+var startableEndpoint = EndpointWithExternallyManagedContainer.Create(endpointConfiguration, serviceCollection);
+
+var containerBuilder = new ContainerBuilder();
+containerBuilder.Populate(serviceCollection);
+var autofacContainer = containerBuilder.Build();
+var endpointInstance = await startableEndpoint.Start(new AutofacServiceProvider(autofacContainer));
+```
+
+#### StructureMap
+
```
Comment (mauroservienti) [inline nservicebus/upgrades/7to8/dependency-injection.md:64]:
```suggestion

Note: StructureMap has been sunsetted. The maintainers recommend to use [Lamar](https://jasperfx.github.io/lamar/) instead.

```

**Review by mauroservienti — CHANGES_REQUESTED** [review]

File: nservicebus/upgrades/7to8/dependency-injection.md (hunk 4 above)
Comment (timbussmann) [reply nservicebus/upgrades/7to8/dependency-injection.md:61]: are you suggesting moving `var containerBuilder = new ContainerBuilder();` up between line 54-55? The remaining order needs to remain I think?

Comment (mauroservienti) [reply nservicebus/upgrades/7to8/dependency-injection.md:61]: Which diminishes the importance. Scratch that. It's OK as it is now.

**Review by mauroservienti — APPROVED** [review]

File: nservicebus/upgrades/7to8/dependency-injection.md [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+The following snippet configures NServiceBus to use Unity as its dependency injection container. This requires the `Castle.Windsor.MsDependencyInjection` NuGet package.
+
+```csharp
+var serviceCollection = new ServiceCollection();
+var startableEndpoint = EndpointWithExternallyManagedContainer.Create(endpointConfiguration, serviceCollection);
+
+var container = new WindsorContainer();
+var serviceProvider = WindsorRegistrationHelper.CreateServiceProvider(container, serviceCollection);
+var endpointInstance = await startableEndpoint.Start(serviceProvider);
+```
```
Comment (adamralph) [inline nservicebus/upgrades/7to8/dependency-injection.md:103]: @timbussmann @mauroservienti the references to `Castle.Windsor` look like a copy paste error?

Comment (timbussmann) [reply nservicebus/upgrades/7to8/dependency-injection.md:103]: um, yeah it does 🤦

Comment (timbussmann) [reply nservicebus/upgrades/7to8/dependency-injection.md:103]: raised https://github.com/Particular/docs.particular.net/pull/5304

**Pull request merged**